swantbls/swantbls.6
.TH swantbls 6 "13 January 2000"

.SH NAME
swantbls \- create lumps for customized switches and animated textures and flats for inclusion in PWAD files

.SH SYNOPSIS
.B swantbls
.I infile

.SH DISCLAIMER
This program is furnished on an AS IS basis for the fun of it.
id Software does not support this program.

.SH DESCRIPTION
The program takes a text input file and creates binary lump files
switches.lmp and animated.lmp for inclusion in PWAD files. This
allows PWAD authors to customize switches and animations for some extended
Doom engines.

.SH "INPUT FORMAT"
The input file syntax is as follows:
.LP
.I Switches section:
.nf
[SWITCHES]
<episode>   <texture1>   <texture2>
\|.\|.\|.
.fi
.LP
.I Episode
is a number 1 \- 3 for the IWAD file, in which this switch can be used,
1 = Shareware Doom, 2 = registered Doom, 3 = Doom][.
.LP
.I Texture1
is the name for the texture used, if the switch wasn't activated,
.I texture2
is the name for the texture used, after the switch was activated.
.LP
.I Animated flats section:
.nf
[FLATS]
<number>   <last>   <first>
\|.\|.\|.
.fi
.LP
.I Number
is the number of frames between changes, \fIlast\fR is the name of the
last texture used for the animation, \fIfirst\fR is the name of the
first texture used for the animation.
.LP
.I Animated textures section:
.nf
[TEXTURES]
<number>   <last>   <first>
\|.\|.\|.
.fi
.LP
The meaning of the fields is the same as in the flats section.
.LP
An example file with the definitions for the original Doom games
is included in the WAD tool collection.

.SH "SUPPORTED GAMES"
This data lumps can be generated for any WAD file based engine, but
the engine has to use this data. Engines like \fBBoom\fR, \fBXDoom\fR
and \fBZDoom\fR have support for customized switches and animations.

.SH "SEE ALSO"
.BR mkwad "(6), "
.BR wadgc "(6), "
.BR xwadtools (6)

.SH BUGS/ERRORS
None.

.SH AUTHOR
.B Swantbls
originally was written by Jim Flynn (72713.3334@compuserve.com) for
the Boom engine by TeamTNT.
.LP
Ported to UNIX and fixed to support big endian systems by Udo Munk
(um@compuserve.com).


swantbls/swantbls.c
////////////////////////////////////////////////////////////////////////////
//
// SWANTBLS		By J. Flynn 5/23/98
//
// Converts a text file (.DAT) containing sections defining switches and 
// animated textures and flats into two lumps, SWITCHES.LMP and ANIMATED.LMP,
// suitable for inclusion in a PWAD. See the example text file DEFSWANI.DAT,
// for text file syntax.
//
////////////////////////////////////////////////////////////////////////////

// UM 1999-06-25
// - reformatted a little bit
// - made filenames lower case, DOS/Windows doesn't care, but UNIX does
// - use endian aware functions from the lib to write binary data into files
// - fixed the usual / \ path problem
//
// UM 1999-06-30
// - added string defines for path/extension to sysdep.h, use those to
//   build filenames

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include "sysdep.h"
#include "strfunc.h"

// trim spaces off left of string

void LeftTrim(char *str)
{
	int k, n;

	k = strlen(str);
	if (k)
	{
		n = strspn(str, " \t");
		memmove(str, str+n, k-n+1);
	}
}

// trim spaces off right of string

void RightTrim(char *str)
{
	int k, n;

	n = k = strlen(str);
	if (k)
	{
		n--;
		while (n >= 0 && (str[n] == ' ' || str[n] == '\t'))
			n--;
		str[n+1] = '\0';
	}
}

// trim spaces off both ends of a string

void Trim(char *str)
{
	RightTrim(str);
	LeftTrim(str);
}

// If the extension is missing, add the extension ext
// If the extension is present, leave it be

void SupplyExt(char *str, char *ext)
{
	char *p, *q;

	if (!str || !*str)
		return;

	p = strrchr(str, PATHSEPERATOR);	// last path seperator
	q = strrchr(str, EXTSEPERATOR);		// last extension seperator
	if (p && q && q < p)	// if is a backslash, and the . is before it
        {
		strcat(str, EXTSEPERATOR_STR);
		strcat(str, ext);	// add an extension, since there is none
	}
	else if (!q)
	{
		strcat(str, EXTSEPERATOR_STR);
		strcat(str, ext);	// if no period, add the extension
	}
}

int main(int argc, char **argv)
{
	char inpname[256], *p, *q, *tok;
	char secname[256] = "";
	char string1[9], string2[9];
	char buffer[256];
	int lineno = 0, inum;
	short num;
	char true = 1, false = 0;
	char zero[9] = { 0, 0, 0, 0, 0, 0, 0, 0, 0 };
	FILE *st = NULL, *outst = NULL;

	if (argc < 2)
	{
		printf("\nUsage: swantbl swanfile[.dat]\n");
		printf("Creates switches.lmp and animated.lmp from text input file\n");
		exit(1);
	}

	strcpy(inpname, argv[1]);
	SupplyExt(inpname, "dat");

	st = fopen(inpname, "r");
	if (st)
	{
		while ((p = fgets(buffer, 256, st)))
		{
			lineno++;
			Trim(p);
			if (p[strlen(p)-1] == '\n') p[strlen(p)-1] = '\0';
			if (*p == '[')
			{
				q = strchr(p, ']');
				if (!q)
				{
					printf("Unclosed section delimiter in line %d, missing ]\n", lineno);
					fclose(st);
					exit(1);
				}

				*q = '\0';
				strcpy(secname, p+1);
				if (!stricmp(secname, "SWITCHES"))
				{
					outst = fopen("switches.lmp", "wb");
				}
				else if (!stricmp(secname, "FLATS"))
				{
					;
				}
				else if (!stricmp(secname, "TEXTURES"))
				{
					;
				}
				else 
				{
					printf("Bad or duplicate section header in line %d, use [SWITCHES], [FLATS], or [TEXTURES] once\n", lineno);
					fclose(st);
					exit(1);
				}
			}
			else if (*p == '\n' || *p == '\r' || *p == '#' ||
				 *p == ';' || *p == 0)
			{
				continue;
			}
			else if (!isdigit(*p))
			{
				printf("Bad syntax, line %d does not begin with a number\n", lineno);
				exit(1);
			}
			else if (!stricmp("SWITCHES", secname))
			{
				tok = strtok(p, " \t,");
				if (!tok)
				{
					printf("Bad syntax on line %d, missing first token\n", lineno);
			    		fclose(st);
					fclose(outst);
				  	exit(1);
				}
				else
				{
					num = inum = atoi(tok);
					tok = strtok(NULL, " \t,");
					if (!tok)
					{
						printf("Bad syntax on line %d, missing second token\n", lineno);
			    			fclose(st);
						fclose(outst);
				  		exit(1);
					}
					else
					{
						strncpy(string1, tok, 9);
						tok = strtok(NULL, " \t,");
						if (!tok)
						{
							printf("Bad syntax on line %d, missing third token\n", lineno);
			    				fclose(st);
							fclose(outst);
				  			exit(1);
						}
						else strncpy(string2, tok, 9);
					}
				}
				if (!stricmp(secname, "SWITCHES"))
				{
					fwrite(string1, 9, 1, outst);
					fwrite(string2, 9, 1, outst);
					putshort(num, outst);
				}
				else
				{
					printf("Error: data at line %d found outside section\n", lineno);
					fclose(st);
					fclose(outst);
				}
			}
		}
		if (stricmp("SWITCHES", secname))
		{
			fwrite(zero, 9, 1, outst);
			fwrite(zero, 9, 1, outst);
			fwrite(zero, 2, 1, outst);
		}
		if (outst) fclose(outst);
		outst = NULL;

		fclose(st);
		st = fopen(inpname, "r");
		while (st && (p = fgets(buffer, 256, st)))
		{
			lineno++;
			Trim(p);
			if (p[strlen(p)-1] == '\n') p[strlen(p)-1] = '\0';
			if (*p == '[')
			{
				q = strchr(p, ']');
				if (!q)
				{
					printf("Unclosed section delimiter in line %d, missing ]\n", lineno);
					fclose(st);
					exit(1);
				}

				*q = '\0';
				strcpy(secname, p+1);
				if (!stricmp(secname, "SWITCHES"))
				{
					;
				}
				else if (!stricmp(secname, "FLATS"))
				{
					if (!outst)
						outst = fopen("animated.lmp", "wb");
				}
				else if (!stricmp(secname, "TEXTURES"))
				{
					if (!outst)
						outst = fopen("animated.lmp", "wb");
				}
				else 
				{
					printf("Bad or section header in line %d, use [SWITCHES], [FLATS], or [TEXTURES] once\n", lineno);
					fclose(st);
					exit(1);
				}
			}
			else if (*p == '\n' || *p == '\r' || *p == '#'
				 || *p == ';' || *p == '\0')
			{
				continue;
			}
			else if (!isdigit(*p))
			{
				printf("Bad syntax, line %d does not begin with a number\n", lineno);
				exit(1);
			}
			else if (!stricmp("FLATS", secname) ||
				 !stricmp("TEXTURES", secname))
			{
				tok = strtok(p, " \t,");
				if (!tok)
				{
					printf("Bad syntax on line %d, missing first token\n", lineno);
			    		fclose(st);
					fclose(outst);
					exit(1);
				}
				else
				{
					num = inum = atoi(tok);
					tok = strtok(NULL, " \t,");
					if (!tok)
					{
						printf("Bad syntax on line %d, missing second token\n", lineno);
			    			fclose(st);
						fclose(outst);
				  		exit(1);
					}
					else
					{
						strncpy(string1, tok, 9);
						tok = strtok(NULL, " \t,");
						if (!tok)
						{
							printf("Bad syntax on line %d, missing third token\n", lineno);
			    				fclose(st);
							fclose(outst);
				  			exit(1);
						}
						else strncpy(string2, tok, 9);
					}
				}
				if (!stricmp(secname, "FLATS"))
				{
					fwrite(&false, 1, 1, outst);
					fwrite(string1, 9, 1, outst);
					fwrite(string2, 9, 1, outst);
					putlong(inum, outst);
				}
				else if (!stricmp(secname, "TEXTURES"))
				{
					fwrite(&true, 1, 1, outst);
					fwrite(string1, 9, 1, outst);
					fwrite(string2, 9, 1, outst);
					putlong(inum, outst);
				}
				else
				{
					printf("Error: data at line %d found outside section\n", lineno);
					fclose(st);
					fclose(outst);
					exit(1);
				}
			}
		}
		if (!stricmp(secname, "FLATS"))
		{
			fwrite("\xff\xff\xff\xff", 4, 1, outst);
		}
		else if (!stricmp(secname, "TEXTURES"))
		{
			fwrite("\xff\xff\xff\xff", 4, 1, outst);
		}
	}
	else
	{
		printf("Cannot open %s for reading\n", inpname);
		exit(1);
	}

	if (st)
		fclose(st);
	if (outst)
		fclose(outst);

	return 0;
}
